hw/ip/otbn/dv/rig/rig/gens/bad_load_store.py

```diff
@@ -0,0 +1,284 @@
+            if weight > 0:
+                self.weights.append(weight)
+                self.insns.append(insn)
+
+        # Check that at least one instruction has a positive weight
+        assert len(self.insns) == len(self.weights)
+        if not self.weights:
+            self.disabled = True
+
+    def gen(self,
+            cont: GenCont,
+            model: Model,
+            program: Program) -> Optional[GenRet]:
+
+        weights = self.weights
+        prog_insn = None
+        while prog_insn is None:
+            idx = random.choices(range(len(self.insns)), weights=weights)[0]
+            # At least one weight should be positive. This is guaranteed so
+            # long as fill_insn doesn't fail for absolutely everything. We know
+            # that doesn't happen, because we have some instructions (such as
+            # addi) where it will not fail.
+            assert weights[idx] > 0
+
+            # Try to fill out the instruction. On failure, clear the weight for
+            # this index and go around again. We take the copy here, rather
+            # than outside the loop, because we don't expect this to happen
+            # very often.
+            prog_insn = self.fill_insn(self.insns[idx], model)
+            if prog_insn is None:
+                weights = self.weights.copy()
+                weights[idx] = 0
+                continue
+
+        snippet = ProgSnippet(model.pc, [prog_insn])
+        snippet.insert_into_program(program)
+
+        return (snippet, True, model)
+
+    def fill_insn(self, insn: Insn, model: Model) -> Optional[ProgInsn]:
+        '''Try to pick one of BN.XID or XW instructions
+
+        '''
+
+        # Special-case BN load/store instructions by mnemonic. These use
+        # complicated indirect addressing, so it's probably more sensible to
+        # give them special code.
+        if insn.mnemonic in ['bn.lid', 'bn.sid']:
+            return self._fill_bn_xid(insn, model)
+        if insn.mnemonic in ['lw', 'sw']:
+            return self._fill_xw(insn, model)
+
+        return None
+
+    def _fill_xw(self, insn: Insn, model: Model) -> Optional[ProgInsn]:
+
+        # LW has three operands: grd, offset, grs
+        if not (len(self.lw.operands) == 3 and
+                isinstance(self.lw.operands[0].op_type, RegOperandType) and
+                self.lw.operands[0].op_type.reg_type == 'gpr' and
+                self.lw.operands[0].op_type.is_dest() and
+                isinstance(self.lw.operands[1].op_type, ImmOperandType) and
+                isinstance(self.lw.operands[2].op_type, RegOperandType) and
+                self.lw.operands[2].op_type.reg_type == 'gpr' and
+                not self.lw.operands[2].op_type.is_dest()):
+            raise RuntimeError('LW instruction from instructions file is '
+                               'not the shape expected by the BadLoadStore'
+                               'generator.')
```

Comment:
I'd probably put this check into the constructor. No need to check each time we fill an instruction, I don't think. Similarly for SW, BN.LID and BN.SID.